pkg/api/norman/customization/aks/listers.go

```diff
@@ -0,0 +1,192 @@
+package aks
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"regexp"
+	"sort"
+
+	"github.com/Azure/azure-sdk-for-go/services/containerservice/mgmt/2020-09-01/containerservice"
+	"github.com/Azure/azure-sdk-for-go/services/network/mgmt/2020-07-01/network"
+	"github.com/Azure/go-autorest/autorest"
+	"github.com/Azure/go-autorest/autorest/adal"
+	"github.com/mcuadros/go-version"
+)
+
+type virtualNetworksResponseBody struct {
+	Name          string   `json:"name"`
+	ResourceGroup string   `json:"resourceGroup"`
+	Subnets       []subnet `json:"subnets"`
+}
+
+type subnet struct {
+	Name         string `json:"name"`
+	AddressRange string `json:"addressRange"`
+}
+
+var matchAzureNetwork = regexp.MustCompile("/resourceGroups/(.+?)/")
+
+func NewClientAuthorizer(cap *Capabilities) (autorest.Authorizer, error) {
+	oauthConfig, err := adal.NewOAuthConfig(cap.AuthBaseURL, cap.TenantID)
+	if err != nil {
+		return nil, err
+	}
+
+	spToken, err := adal.NewServicePrincipalToken(*oauthConfig, cap.ClientID, cap.ClientSecret, cap.BaseURL)
+	if err != nil {
+		return nil, fmt.Errorf("couldn't authenticate to Azure cloud with error: %v", err)
+	}
+
+	return autorest.NewBearerAuthorizer(spToken), nil
+}
+
+func NewContainerServiceClient(cap *Capabilities) (*containerservice.ContainerServicesClient, error) {
+	authorizer, err := NewClientAuthorizer(cap)
+	if err != nil {
+		return nil, err
+	}
+
+	containerService := containerservice.NewContainerServicesClientWithBaseURI(cap.BaseURL, cap.SubscriptionID)
+	containerService.Authorizer = authorizer
+
+	return &containerService, nil
+}
+
+func NewNetworkServiceClient(cap *Capabilities) (*network.VirtualNetworksClient, error) {
+	authorizer, err := NewClientAuthorizer(cap)
+	if err != nil {
+		return nil, err
+	}
+
+	containerService := network.NewVirtualNetworksClientWithBaseURI(cap.BaseURL, cap.SubscriptionID)
+	containerService.Authorizer = authorizer
+
+	return &containerService, nil
+}
+
+type sortableVersion []string
+
+func (s sortableVersion) Len() int {
+	return len(s)
+}
+
+func (s sortableVersion) Swap(a, b int) {
+	s[a], s[b] = s[b], s[a]
+}
+
+func (s sortableVersion) Less(a, b int) bool {
+	return version.Compare(s[a], s[b], "<")
+}
+
+func listKubernetesVersions(ctx context.Context, cap *Capabilities) ([]byte, int, error) {
+	if cap.TenantID == "" || cap.ResourceLocation == "" {
+		return nil, http.StatusBadRequest, fmt.Errorf("tenantId and region are required")
+	}
```

Comment:
Why not put checks that belon with NewContainerServiceClient into those methods ? Meaning tenantId, as that requirement seems to be in that method, and only resourceLocation is required here. Same for other check below with NewNetworkServiceClient.

Reply:
Updated

Reply:
Check on tenantID was removed as its required earlier in the cloud-credential process.

Reply:
I don't see where this was moved to

Reply:
I guess it's not needed for this function